README.md
<p align="center"><img src="https://cdn.jsdelivr.net/gh/tagproject/art/packages/stdout-update/banner.svg" alt="Package logo"></p>

<p align="center">
    <a href="https://travis-ci.com/keindev/stdout-update"><img src="https://travis-ci.com/keindev/stdout-update.svg?branch=master" alt="Build Status"></a>
    <a href="https://codecov.io/gh/keindev/stdout-update"><img src="https://codecov.io/gh/keindev/stdout-update/branch/master/graph/badge.svg" /></a>
    <a href="https://www.npmjs.com/package/stdout-update"><img alt="npm" src="https://img.shields.io/npm/v/stdout-update.svg"></a>
    <a href="https://github.com/tagproject/ts-package-shared-config"><img src="https://img.shields.io/badge/standard--shared--config-nodejs%2Bts-green?logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAfCAYAAACh+E5kAAAACXBIWXMAAAsTAAALEwEAmpwYAAAAAXNSR0IArs4c6QAAAARnQU1BAACxjwv8YQUAAAJQSURBVHgB1VftUcMwDFU4/tMNyAZ0A7IBbBA2CExAmIBjApcJChO0TFA2SJkgMIGRyDNV3TSt26RN353OX/LHUyTZIdoB1tqMZcaS0imBDzxkeWaJWR51SX0HrJ6pdsJyifpdb4loq3v9A+1CaBuWMR0Q502DzuJRFD34Y9z3DXIRNy/QPWKZY27COlM6BtZZHWMJ3CkVa28KZMTJkDpCVLOhs/oL2gMuEhYpxeenPPah9EdczLkvpwZgnQHWnlNLiNQGYiWx5gu6Ehz4m+WNN/2i9Yd75CJmeRDXogbIFxECrqQ2wIvlLBOXaViuYbGQNSQLFSGZyOnulb2wadaGnyoSSeC8GBJkNDf5kloESAhy2gFIIPG2+ufUMtivn/gAEi+Gy4u6FLxh/qer8/xbLq7QlNh6X4mbtr+A3pylDI0Lb43YrmLmXP5v3a4I4ABDRSI4xjB/ghveoj4BCVm37JQADhGDgOA+YJ48TSaoOwKpt27aOQG1WRES3La65WPU3dysTjE8de0Aj8SsKS5sdS9lqCeYI08bU6d8EALYS5OoDW4c3qi2gf7f+4yODfj2DIcqdVzYKnMtEUO7RP2gT/W1AImxXSC3i7R7rfRuMT5G2xzSYzaCDzOyyzDeuNHZx1a3fOdJJwh28fRwwT1QY6Xzf7TvWG6ob/BIGPQ59ymUngRyRn2El6Fy5T7G0zl+JmoC3KRQXyT1xpfiJKIeAemzqBl6U3V5ocZNf4hHg61u223wn4nOqF8IzvF9IxCMkyfQ+i/lnnhlmW6h9+Mqv1SmQhehji4JAAAAAElFTkSuQmCC" alt="Standard Shared Config"></a>
</p>

Purely and accurately overwrites the previous output in the terminal, while maintaining the history of third-party logs.

<p align="center">
    <img src="media/demo.gif">
</p>

## Install

```
npm install stdout-update
```

## Usage

```typescript
import UpdateManager from 'stdout-update';

const TICKS = 60;
const TIMEOUT = 80;
const manager = UpdateManager.getInstance();
const frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
const messages = ['Swapping time and space...', 'Have a good day.', "Don't panic...", 'Updating Updater...', '42'];
let ticks = TICKS;
let i = 0;

manager.hook();

// eslint-disable-next-line no-console
console.log(' - log message');
// eslint-disable-next-line no-console
console.error(' - error message');
// eslint-disable-next-line no-console
console.warn(' - warn message');

const id = setInterval(() => {
  if (--ticks < 0) {
    clearInterval(id);
    manager.update(['✔ Success', '', 'Messages:', 'this line is be deleted!!!']);
    manager.erase(1);
    manager.unhook(false);
  } else {
    const frame = frames[(i = ++i % frames.length)];
    const index = Math.round(ticks / 10) % messages.length;
    const message = messages[index];

    if (message) manager.update([`${frame} Some process...`, message]);
  }
}, TIMEOUT);
```

## Examples

[tasktree-cli](https://github.com/keindev/tasktree) - simple terminal task tree, helps you keep track of your tasks in a tree structure.

## API

Read the [API documentation](docs/api/index.md) for more information.


CHANGELOG.md
# Important Changes

## Engines

- Changed **node** from `^14.13.1 || >=16.0.0` to `>=16.0.0`

## Dependencies

<details>
<summary>Dependencies</summary>

- Changed **[ansi-escapes](https://www.npmjs.com/package/ansi-escapes)** from `^6.0.0` to `^6.2.0`
- Changed **[strip-ansi](https://www.npmjs.com/package/strip-ansi)** from `^7.0.1` to `^7.1.0`
- Bumped **[string-width](https://www.npmjs.com/package/string-width)** from `^5.1.2` to `^6.1.0`

</details>

<details>
<summary>Dev Dependencies</summary>

- Bumped **[@tagproject/ts-package-shared-config](https://www.npmjs.com/package/@tagproject/ts-package-shared-config)** from `^10.0.2` to `^11.0.1`
- Removed **[@tagproject/docs-shared-config](https://www.npmjs.com/package/@tagproject/docs-shared-config)**, with `^1.1.2`
- Removed **[@tagproject/vscode-shared-config](https://www.npmjs.com/package/@tagproject/vscode-shared-config)**, with `^2.0.5`
- Removed **[@types/jest](https://www.npmjs.com/package/@types/jest)**, with `^29.2.3`
- Removed **[@types/node](https://www.npmjs.com/package/@types/node)**, with `^18.11.9`
- Removed **[@typescript-eslint/eslint-plugin](https://www.npmjs.com/package/@typescript-eslint/eslint-plugin)**, with `^5.44.0`
- Removed **[@typescript-eslint/parser](https://www.npmjs.com/package/@typescript-eslint/parser)**, with `^5.44.0`
- Removed **[changelog-guru](https://www.npmjs.com/package/changelog-guru)**, with `^4.0.9`
- Removed **[cspell](https://www.npmjs.com/package/cspell)**, with `^6.15.0`
- Removed **[eslint](https://www.npmjs.com/package/eslint)**, with `^8.28.0`
- Removed **[eslint-config-prettier](https://www.npmjs.com/package/eslint-config-prettier)**, with `^8.5.0`
- Removed **[eslint-plugin-import](https://www.npmjs.com/package/eslint-plugin-import)**, with `^2.26.0`
- Removed **[eslint-plugin-jest](https://www.npmjs.com/package/eslint-plugin-jest)**, with `^27.1.6`
- Removed **[eslint-plugin-node](https://www.npmjs.com/package/eslint-plugin-node)**, with `^11.1.0`
- Removed **[eslint-plugin-optimize-regex](https://www.npmjs.com/package/eslint-plugin-optimize-regex)**, with `^1.2.1`
- Removed **[eslint-plugin-promise](https://www.npmjs.com/package/eslint-plugin-promise)**, with `^6.1.1`
- Removed **[ghinfo](https://www.npmjs.com/package/ghinfo)**, with `^3.0.8`
- Removed **[husky](https://www.npmjs.com/package/husky)**, with `^8.0.2`
- Removed **[jest](https://www.npmjs.com/package/jest)**, with `^29.3.1`
- Removed **[npm-run-all](https://www.npmjs.com/package/npm-run-all)**, with `^4.1.5`
- Removed **[prettier](https://www.npmjs.com/package/prettier)**, with `^2.8.0`
- Removed **[rimraf](https://www.npmjs.com/package/rimraf)**, with `^3.0.2`
- Removed **[ts-jest](https://www.npmjs.com/package/ts-jest)**, with `^29.0.3`
- Removed **[typescript](https://www.npmjs.com/package/typescript)**, with `^4.9.3`

</details>

---

# Contributors

[![@github-actions[bot]](https://avatars.githubusercontent.com/in/15368?v=4&s=40)](https://github.com/github-actions%5Bbot%5D) [![@keindev](https://avatars.githubusercontent.com/u/4527292?v=4&s=40)](https://github.com/keindev)


example/app.ts
/* eslint-disable */
import UpdateManager from '../src/UpdateManager';

const TICKS = 60;
const TIMEOUT = 80;
const manager = UpdateManager.getInstance();
const frames = ['⠋', '⠙', '⠹', '⠸', '⠼', '⠴', '⠦', '⠧', '⠇', '⠏'];
const messages = ['Swapping time and space...', 'Have a good day.', "Don't panic...", 'Updating Updater...', '42'];
let ticks = TICKS;
let i = 0;

manager.hook();

// eslint-disable-next-line no-console
console.log(' - log message');
// eslint-disable-next-line no-console
console.error(' - error message');
// eslint-disable-next-line no-console
console.warn(' - warn message');

const id = setInterval(() => {
  if (--ticks < 0) {
    clearInterval(id);
    manager.update(['✔ Success', '', 'Messages:', 'this line is be deleted!!!']);
    manager.erase(1);
    manager.unhook(false);
  } else {
    const frame = frames[(i = ++i % frames.length)];
    const index = Math.round(ticks / 10) % messages.length;
    const message = messages[index];

    if (message) manager.update([`${frame} Some process...`, message]);
  }
}, TIMEOUT);
